File: Difficulty: Medium/Deletion and Reverse in Circular Linked List/deletion-and-reverse-in-circular-linked-list.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

struct Node {
    int data;
    struct Node *next;

    Node(int x) {
        data = x;
        next = NULL;
    }
};


// } Driver Code Ends
class Solution {
  public:
    /* structure for a node
    struct Node
    {
        int data;
        struct Node *next;
    };
    */

    /* Function to delete a given node from the list */
    void deleteNode(struct Node** head, int key)
    {
          if (*head == NULL) return;
        struct Node* curr = (*head)->next;
        struct Node* prev = *head;
        
        if((*head)->data ==key)
        {
            while(curr->next!=*head) curr = curr->next;
            
            curr->next = (*head)->next;
            delete(*head);
            *head = curr->next;
            return;
        }
        
        do{
            if(curr->data ==key)
            {
                prev->next = curr->next;
                delete(curr);
                return;
            }
            
            prev = curr;
            curr = curr->next;
        }while(curr != *head);
        
       
    }

    /* Function to reverse the linked list */
    void reverse(struct Node** head_ref)
    {
           struct Node* curr = *head_ref;
        struct Node* prev = NULL;
        struct Node* forward = NULL;
        
        do{
            forward = curr->next;
            curr->next= prev;
            prev = curr;
            curr = forward;
        }while(curr != *head_ref);
        
        (*head_ref)->next = prev;
        *head_ref = prev;

    }
};

//{ Driver Code Starts.

/* Function to print nodes in a given
circular linked list */
void printList(struct Node *head) {
    struct Node *temp = head;
    if (head != NULL) {
        do {
            printf("%d ", temp->data);
            temp = temp->next;
        } while (temp != head);
    } else
        cout << "empty" << endl;

    printf("\n");
}

int main() {
    int t;
    cin >> t;
    cin.ignore();
    while (t--) {
        vector<int> arr;
        string input;
        getline(cin, input);
        stringstream ss(input);
        int number;
        while (ss >> number) {
            arr.push_back(number);
        }
        int key;
        cin >> key;
        cin.ignore();
        struct Node *head = new Node(arr[0]);
        struct Node *tail = head;
        for (int i = 1; i < arr.size(); ++i) {
            tail->next = new Node(arr[i]);
            tail = tail->next;
        }
        tail->next = head; // Make the list circular
        Solution ob;
        ob.deleteNode(&head, key);
        ob.reverse(&head);
        printList(head);
    }
    return 0;
}
// } Driver Code Ends

File: Difficulty: Medium/Deletion and Reverse in Circular Linked List/deletion-and-reverse-in-circular-linked-list.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;

class Node {
    int data;
    Node next;

    Node(int x) {
        data = x;
        next = null;
    }
}

class LinkedList {
    // Function to print nodes in a given circular linked list
    static void printList(Node head) {
        if (head == null) {
            System.out.println("empty");
            return;
        }
        Node temp = head;
        do {
            System.out.print(temp.data + " ");
            temp = temp.next;
        } while (temp != head);
        System.out.println();
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());

        while (t-- > 0) {
            List<Integer> arr = new ArrayList<>();
            String input = br.readLine();
            StringTokenizer st = new StringTokenizer(input);
            while (st.hasMoreTokens()) {
                arr.add(Integer.parseInt(st.nextToken()));
            }
            int key = Integer.parseInt(br.readLine());
            Node head = new Node(arr.get(0));
            Node tail = head;
            for (int i = 1; i < arr.size(); ++i) {
                tail.next = new Node(arr.get(i));
                tail = tail.next;
            }
            tail.next = head; // Make the list circular
            Solution ob = new Solution();
            Node current = ob.deleteNode(head, key);
            Node rev = ob.reverse(current);
            printList(rev);
        }
    }
}

// } Driver Code Ends


/*class Node
{
    int data;
    Node next;
    Node(int d)
    {
        data=d;next=null;
    }
}*/

class Solution {
     Node reverse(Node head) {
        // code here
        if(head==null){
            return head;
        }
        Node prev=null;
        Node cur=head;
        Node nex=head;
        
        do{
            nex=cur.next;
            cur.next=prev;
            prev=cur;
            cur=nex;
        }while(cur!=head);
        
        head.next=prev;
        return head=prev;
    }

    // Function to delete a node from the circular linked list
    Node deleteNode(Node head, int key) {
        // code here
        if(head==null){
            return head;
        }
        // if(head.data==key){
        //     return head=null;
        // }
        Node prev=null;
        Node cur=head;
        
        do{
            if(cur.data==key){
                if(prev==null){
                    Node temp=head;
                    do{
                        temp=temp.next;
                    }while(temp.next!=head);
                    temp.next=head.next;
                    return head=head.next;
                }else{
                    prev.next=cur.next;
                    return head;
                }
            }
            prev=cur;
            cur=cur.next;
        }while(cur!=head);
        return head;
    }



}